### ISSUE TITLE
False positive in F541 (f-string without placeholders) after v0.0.204

### ISSUE THREAD

**Issue by tlambert03** [charliermarsh/ruff#1514]
not sure if it's #1494 or a different PR... but a F541 "f-string without placeholders" false positive is now appearing in v0.0.204 when using format specifiers in f-strings:

```python
v = 23.234234
f"{v:0.2f}"  # error
```

**Comment by charliermarsh** [comment]
Thank you for filing.

**Comment by charliermarsh** [comment]
@harupy - Do you have any thoughts on the correct fix here?

The AST looks like this:

```
Expr {
    value: Located {
        location: Location {
            row: 2,
            column: 0,
        },
        end_location: Some(
            Location {
                row: 2,
                column: 11,
            },
        ),
        custom: (),
        node: JoinedStr {
            values: [
                Located {
                    location: Location {
                        row: 2,
                        column: 0,
                    },
                    end_location: Some(
                        Location {
                            row: 2,
                            column: 11,
                        },
                    ),
                    custom: (),
                    node: FormattedValue {
                        value: Located {
                            location: Location {
                                row: 2,
                                column: 3,
                            },
                            end_location: Some(
                                Location {
                                    row: 2,
                                    column: 4,
                                },
                            ),
                            custom: (),
                            node: Name {
                                id: "v",
                                ctx: Load,
                            },
                        },
                        conversion: 0,
                        format_spec: Some(
                            Located {
                                location: Location {
                                    row: 2,
                                    column: 0,
                                },
                                end_location: Some(
                                    Location {
                                        row: 2,
                                        column: 11,
                                    },
                                ),
                                custom: (),
                                node: JoinedStr {
                                    values: [
                                        Located {
                                            location: Location {
                                                row: 2,
                                                column: 0,
                                            },
                                            end_location: Some(
                                                Location {
                                                    row: 2,
                                                    column: 11,
                                                },
                                            ),
                                            custom: (),
                                            node: Constant {
                                                value: Str(
                                                    "0.2f",
                                                ),
                                                kind: None,
                                            },
                                        },
                                    ],
                                },
                            },
                        ),
                    },
                },
            ],
        },
    }
}
```

So the format spec itself is an "empty" `JoinedStr`.

**Comment by charliermarsh** [comment]
I was gonna say we could ignore `format_spec` but... this may require a fix in RustPython, since we _do_ want to flag the latter case:

```
v = 23.234234
f"{v:0.2f}"
f"{v:{f'0.2f'}}"
```

**Comment by AA-Turner** [comment]
Another reproducer:

```
(sphinx) PS S:\Development\sphinx> type bug.py                
f' style="vertical-align: {-0:d}px"'
(sphinx) PS S:\Development\sphinx> pip install "ruff==0.0.203" --quiet
(sphinx) PS S:\Development\sphinx> ruff -V
ruff 0.0.203
(sphinx) PS S:\Development\sphinx> ruff --select F bug.py             
(sphinx) PS S:\Development\sphinx> pip install "ruff==0.0.204" --quiet
(sphinx) PS S:\Development\sphinx> ruff -V                            
ruff 0.0.204
(sphinx) PS S:\Development\sphinx> ruff --select F bug.py             
bug.py:1:1: F541 f-string without any placeholders
Found 1 error(s).
```

A

**Comment by charliermarsh** [comment]
I can probably get something out quick-ish to resolve these cases (at the cost of some rare false-negatives).

**Comment by charliermarsh** [comment]
This does feel like a RustPython bug though. \cc @harupy just for visibility.

**Comment by harupy** [comment]
Checked python's AST for `f"{v:0.2}"`:

```
>>> print(ast.dump(ast.parse('f"{v:0.2}"'), include_attributes=True, indent=2))
Module(
  body=[
    Expr(
      value=JoinedStr(
        values=[
          FormattedValue(
            value=Name(
              id='v',
              ctx=Load(),
              lineno=1,
              col_offset=3,
              end_lineno=1,
              end_col_offset=4),
            conversion=-1,
            format_spec=JoinedStr(
              values=[
                Constant(
                  value='0.2',
                  lineno=1,
                  col_offset=0,
                  end_lineno=1,
                  end_col_offset=10)],
              lineno=1,
              col_offset=0,
              end_lineno=1,
              end_col_offset=10),
            lineno=1,
            col_offset=0,
            end_lineno=1,
            end_col_offset=10)],
        lineno=1,
        col_offset=0,
        end_lineno=1,
        end_col_offset=10),
      lineno=1,
      col_offset=0,
      end_lineno=1,
      end_col_offset=10)],
  type_ignores=[])
```

**Comment by charliermarsh** [comment]
Oh nice, so it's the same?

**Comment by harupy** [comment]
It looks the same.

**Comment by charliermarsh** [comment]
It looks like Pyflakes doesn't check nested f-strings:

```py
_in_fstring = False

def JOINEDSTR(self, node):
    if (
            # the conversion / etc. flags are parsed as f-strings without
            # placeholders
            not self._in_fstring and
            not any(isinstance(x, ast.FormattedValue) for x in node.values)
    ):
        self.report(messages.FStringMissingPlaceholders, node)

    self._in_fstring, orig = True, self._in_fstring
    try:
        self.handleChildren(node)
    finally:
        self._in_fstring = orig
```

**Comment by charliermarsh** [comment]
Probably the best we can do then? I can fix.

**Comment by charliermarsh** [comment]
I did suspect that `0.2` part would at least have a different location though. It looks like it uses the same location as the surrounding f-string. If it had a different location, I could extract the raw text and manually confirm that it's not a "real" f-string.

**Comment by tlambert03** [comment]
thanks as always for the blazing fast response

**Comment by harupy** [comment]
My attempt to fix this issue:

```diff
diff --git a/src/checkers/ast.rs b/src/checkers/ast.rs
index 8629bae..0079d8b 100644
--- a/src/checkers/ast.rs
+++ b/src/checkers/ast.rs
@@ -2151,7 +2151,20 @@ where
                 }
             }
             ExprKind::JoinedStr { values } => {
-                if self.settings.enabled.contains(&CheckCode::F541) {
+                let is_format_spec = match self.current_expr_parent() {
+                    Some(parent) => match &parent.0.node {
+                        ExprKind::FormattedValue { format_spec, .. } => match format_spec {
+                            Some(format_spec) => {
+                                format_spec.location == expr.location
+                                    && format_spec.end_location == expr.end_location
+                            }
+                            None => false,
+                        },
+                        _ => false,
+                    },
+                    None => false,
+                };
+                if !is_format_spec && self.settings.enabled.contains(&CheckCode::F541) {
                     if !values
                         .iter()
                         .any(|value| matches!(value.node, ExprKind::FormattedValue { .. }))
```

This yields:

```
> cargo run -- --no-cache a.py --show-source
   Compiling ruff v0.0.204 (/home/haru/Desktop/repositories/ruff)
    Finished dev [unoptimized + debuginfo] target(s) in 7.60s
     Running `target/debug/ruff --no-cache a.py --show-source`
a.py:6:5: F541 f-string without any placeholders
  |
6 | c = f"def"
  |     ^^^^^^ F541
  |

a.py:7:5: F541 f-string without any placeholders
  |
7 | d = f"def" + "ghi"
  |     ^^^^^^ F541
  |

a.py:9:5: F541 f-string without any placeholders
  |
9 |     f"def" +
  |     ^^^^^^ F541
  |

a.py:17:5: F541 f-string without any placeholders
   |
17 | h = "x" "y" f"z"
   |     ^^^^^^^^^^^^ F541
   |

a.py:25:7: F541 f-string without any placeholders
   |
25 | f"{v:{f'0.2f'}}"
   |       ^^^^^^^ F541
   |

a.py:26:4: F541 f-string without any placeholders
   |
26 | f"{f''}"
   |    ^^^ F541
   |

Found 6 error(s).
```

**Comment by charliermarsh** [comment]
Oh nice. @harupy do you want to PR that? At your convenience. It's not urgent. That first fix is building now in [v0.0.205](https://github.com/charliermarsh/ruff/releases/tag/v0.0.205) anyway.

**Comment by harupy** [comment]
Yes, I'd be happy to PR that.

**Comment by harupy** [comment]
@charliermarsh I came up with a different approach. The diff looks like this:

```diff
diff --git a/src/ast/visitor.rs b/src/ast/visitor.rs
index f531bba..9bcefd0 100644
--- a/src/ast/visitor.rs
+++ b/src/ast/visitor.rs
@@ -387,7 +387,16 @@ pub fn walk_expr<'a, V: Visitor<'a> + ?Sized>(visitor: &mut V, expr: &'a Expr) {
         ExprKind::FormattedValue {
            value, format_spec, ..
         } => {
             visitor.visit_expr(value);
             if let Some(expr) = format_spec {
-                visitor.visit_expr(expr);
+                match &expr.node {
+                    // Conversion flags are parsed as f-strings without placeholders,
+                    // which would lead to false positives for F541 (FStringMissingPlaceholders).
+                    ExprKind::JoinedStr { values } => {
+                        for expr in values {
+                            visitor.visit_expr(expr);
+                        }
+                    }
+                    _ => unreachable!("Unexpected expression for format_spec"),
+                }
             }
         }
         ExprKind::JoinedStr { values } => {
```

This approach yields:

```
> cargo run -- --no-cache --show-source resources/test/fixtures/pyflakes/F541.py
resources/test/fixtures/pyflakes/F541.py:6:5: F541 f-string without any placeholders
  |
6 | c = f"def"
  |     ^^^^^^ F541
  |

resources/test/fixtures/pyflakes/F541.py:7:5: F541 f-string without any placeholders
  |
7 | d = f"def" + "ghi"
  |     ^^^^^^ F541
  |

resources/test/fixtures/pyflakes/F541.py:9:5: F541 f-string without any placeholders
  |
9 |     f"def" +
  |     ^^^^^^ F541
  |

resources/test/fixtures/pyflakes/F541.py:17:5: F541 f-string without any placeholders
   |
17 | h = "x" "y" f"z"
   |     ^^^^^^^^^^^^ F541
   |

resources/test/fixtures/pyflakes/F541.py:26:7: F541 f-string without any placeholders
   |
26 | f"{v:{f'0.2f'}}"
   |       ^^^^^^^ F541
   |

resources/test/fixtures/pyflakes/F541.py:27:4: F541 f-string without any placeholders
   |
27 | f"{f''}"
   |    ^^^ F541
   |

Found 6 error(s).
```

What do you think?
